openlr/decoded_path.cpp

```diff
@@ -0,0 +1,144 @@
+#include "openlr/decoded_path.hpp"
+
+#include "indexer/index.hpp"
+
+#include "platform/country_file.hpp"
+
+#include "geometry/latlon.hpp"
+#include "geometry/mercator.hpp"
+
+#include "base/assert.hpp"
+#include "base/string_utils.hpp"
+
+namespace
+{
+void LatLonToXML(ms::LatLon const & latLon, pugi::xml_node & node)
+{
+  node.append_child("lat").text() = latLon.lat;
+  node.append_child("lon").text() = latLon.lon;
+}
+
+void LatLonFromXML(pugi::xml_node const & node, ms::LatLon & latLon)
+{
+  latLon.lat = node.child("lat").text().as_double();
+  latLon.lon = node.child("lon").text().as_double();
+}
+
+void FeatureIdFromXML(pugi::xml_node const & node, Index const & index, FeatureID & fid)
+{
+  auto const countryName = node.child("CountryName").text().as_string();
+  fid.m_mwmId = index.GetMwmIdByCountryFile(platform::CountryFile(countryName));
+  fid.m_index = node.child("Index").text().as_uint();
+}
+
+void FeatureIdToXML(FeatureID const & fid, pugi::xml_node & node)
+{
+  node.append_child("CountryName").text() = fid.m_mwmId.GetInfo()->GetCountryName().data();
+  node.append_child("Index").text() = fid.m_index;
+}
+}  // namespace
+
+namespace openlr
+{
+void WriteAsMappingForSpark(std::string const & fileName, std::vector<DecodedPath> const & paths)
+{
+  std::ofstream ost(fileName);
+  if (!ost.is_open())
+    MYTHROW(DecodedPathSaveError, ("Can't write to file", fileName, strerror(errno)));
+
+  WriteAsMappingForSpark(ost, paths);
+
+  if (ost.fail())
+  {
+    MYTHROW(DecodedPathSaveError,
+            ("An error occured while writing file", fileName, strerror(errno)));
+  }
+}
+
+void WriteAsMappingForSpark(std::ostream & ost, std::vector<DecodedPath> const & paths)
+{
+  ost << std::fixed;  // Avoid scientific notation cause '-' is used as fields separator.
+  for (auto const & p : paths)
+  {
+    if (p.m_path.empty())
+      continue;
+
+    ost << p.m_segmentId.Get() << '\t';
+
+    for (auto it = begin(p.m_path); it != end(p.m_path); ++it)
+    {
+      auto const & fid = it->GetFeatureId();
+      ost << fid.m_mwmId.GetInfo()->GetCountryName() << '-'
+          << fid.m_index << '-' << it->GetSegId()
+          << '-' << (it->IsForward() ? "fwd" : "bwd")
+          << '-' << MercatorBounds::DistanceOnEarth(GetStart(*it), GetEnd(*it));
+
+      if (next(it) != end(p.m_path))
+        ost << '=';
+    }
+    ost << endl;
+  }
+}
+
+void PathFromXML(pugi::xml_node const & node, Index const & index, Path & p)
+{
+  auto const edges = node.select_nodes("RoadEdge");
+  for (auto const xmlE: edges)
+  {
+    auto e = xmlE.node();
+
+    FeatureID fid;
+    FeatureIdFromXML(e.child("FeatureID"), index, fid);
```

Comment:
Ok. Although I find it strange that your IntegerFromXML in this PR returns bool instead of throwing.
Consider unifying the error handling style of these xml helpers.